Improve commands update and convert missing commands

fuji97 opened the pull request:
I tried to run the current version of this branch to try the new feature, but I had a lot of problem during startup caused by the bot trying to set the commands when there is no guild with the bot. So I tried to extend it to make it more usable. The following changes were made:

- During startup, the bot iterate to all guilds and try to update the commands in all of them. This should be replaced in the final version with a more flexible solution (like a command to update all the commands in all the guilds).
- Commands are also set during the guild creation phase, so there is no need to restart the bot to update the commands when added to a new guild.
- Refactor `remove` command to work as slash command. Parameters has been changed to position and number of songs to delete, I though this was the best combination fto use with the slash command system. If you know a better solution to this, let me know and I'll update the command.
- Changed `lint` and `lint:fix` npm commands to work in Windows, this should not affects Linux or MacOS.

Also, I'm not that good with Node.js, TypeScript and Discord.js, so I don't know if these are the best way to implement this changes, let me know if something can be done better and I'll update it.

Also, tell me if I should exclude the new yarn.lock, probably yes I think.

codetheweb commented:
Thanks for helping out, I appreciate it. Turned out to be a bigger undertaking than I was expecting. :)

What version of Yarn are you on? It looks like the lockfile's format just slightly changed so not sure what's going on there.

codetheweb commented on src/events/guild-create.ts:
```diff
@@ -1,14 +1,25 @@
   await Settings.upsert({guildId: guild.id, prefix: DEFAULT_PREFIX});
 
+  // Setup slash commands
+  const commands: ApplicationCommandData[] = container.getAll<Command>(TYPES.Command)
```
As per a note in the draft PR, I was planning on registering commands bot-wide in production mode and guild-wide in development mode.

So I think this needs to be wrapped in a check to see if we're in dev mode, and if you have the time it'd be great if you could add bot-wide registering in bot.ts for production mode.

You can check the current environment with `process.env.NODE_ENV===production` (probably should be added to `services/config.ts`). The Dockerfile will also need to be updated to set the environment to production, and it'd be nice if Muse said whether it was registering bot-wide or guild-wide commands when starting up.

fuji97 replied:
Ok, I'm going to change it as soon as I can.
I also have to look up how to set commands bot-wide, do you already have looked up how to do it?

codetheweb replied:
https://discordjs.guide/interactions/registering-slash-commands.html#global-commands

fuji97 commented:
Ok, I've changed the commands setup system and updated the README to setup NODE_ENV.
I've also checked on what version of yarn I am and I'm currently on the v1.22.15 on Windows

codetheweb commented on README.md:
````diff
@@ -48,7 +48,7 @@ There are a variety of image tags available:
 
 ```bash
-docker run -it -v "$(pwd)/data":/data -e DISCORD_TOKEN='' -e SPOTIFY_CLIENT_ID='' -e SPOTIFY_CLIENT_SECRET='' -e YOUTUBE_API_KEY='' codetheweb/muse:latest
+docker run -it -v "$(pwd)/data":/data -e DISCORD_TOKEN='' -e SPOTIFY_CLIENT_ID='' -e SPOTIFY_CLIENT_SECRET='' -e YOUTUBE_API_KEY='' -e NODE_ENV='development' codetheweb/muse:latest
````
Shouldn't this be `NODE_ENV=production` instead?

We should set this directly in the Dockerfile instead of making the consumer do it.

fuji97 replied:
Ok, I moved the variable directly  in the Dockerfile, but I think it's better to default to develop and let the user specify when he wants to deploy in production? I've commented the NODE_ENV override in docker-compose so they can just uncomment it to set in production. I don't know in Node, but in C# the projects default to develop, maybe it's different in Node...
Let me know what's better

codetheweb replied:
Hmm, I think you might be right.

The main downside of global commands is that it can take up to an hour for updates to propagate, so keep `NODE_ENV` defaulting to empty / development for now I guess and users can switch to production if they happen to be running the bot in a very large number of guilds for some reason.

codetheweb commented on src/bot.ts:
```diff
@@ -109,9 +114,32 @@ export default class {
+      // Update commands
+      const rest = new REST({version: '9'}).setToken(this.token);
+
+      switch (this.env) {
```
A simple if-statement would probably be cleaner here since there's only 2 cases.

codetheweb commented on src/services/config.ts:
```diff
@@ -23,6 +24,7 @@ export default class Config {
   readonly YOUTUBE_API_KEY!: string;
   readonly SPOTIFY_CLIENT_ID!: string;
   readonly SPOTIFY_CLIENT_SECRET!: string;
+  readonly NODE_ENV!: string;
```
We only have two modes, development and production. I think it might make more sense to have `IS_PRODUCTION = process.env.NODE_ENV === 'production'` here instead, so we don't have to manually compare the value with a string elsewhere.

codetheweb reviewed (commented):
Thanks so much for your contributions, looking great. 😄

codetheweb commented on src/bot.ts:
```diff
@@ -109,9 +118,29 @@ export default class {
+      } else {
+        // If development, set commands guild-wide
+        this.client.guilds.cache.each(async guild => {
+          await rest.put(
```
I think you probably want to wrap this in a `Promise.all(...)` (using `.cache.map` instead of `.cache.each`) so we wait for all the API calls to succeed. I don't think it's currently waiting for Promises to resolve.

fuji97 replied:
Done, check if it is correct

codetheweb replied:
Promises are now native so you don't need bluebird unless you're doing something fancy. :)

Updated.

codetheweb commented on src/events/guild-create.ts:
```diff
@@ -1,14 +1,30 @@
+  const config = container.get<Config>(TYPES.Config);
+
+  // Setup slash commands
+  if (config.IS_PRODUCTION) {
```
```suggestion
  if (!config.IS_PRODUCTION) {
```

I think? If I'm understanding correctly.

fuji97 replied:
Well, this was a bad typo, corrected

codetheweb commented:
Something funky happened because I did not mean to merge that right away but it merged as soon as I pushed. I think the GH CLI tool got confused between branches. 

In any case, I'll work on clearing out the remaining tasks in https://github.com/codetheweb/muse/pull/431. No promises on a timeline but it's the main thing I'll be working on in regards to Muse.
